### ISSUE TITLE
Next steps for Polonium

### ISSUE THREAD

**Issue by zeroxoneafour** [zeroxoneafour/polonium#88]
I've left this project on the backburner for quite a while, and I'm working on finding a way to improve it. In my opinion, it works just fine right now. However, the code is very messy and does not really reflect what a true tiling manager should be. Polonium came from my previous script, kwin-autotile, and a lot of the very poor programming practices that I had when creating autotile were inherited by Polonium. The code has become increasingly unmaintainable as new features were patched on top of existing code that was unwilling to adapt to them. This is why I have decided to take a step back...

... and rewrite the project from scratch.

I aim to get a beta for Polonium v1.0.0 released by October, with a massively improved and more readable and expandable code base. The new Polonium will be much more abstract in the code, and things like minimum window sizes and other important details will be placed in the center of the engine where they belong, instead of patched on. Engines themselves will be very abstract, and only reflect an idea of how the windows will be tiled. This should lead to many fewer engine bugs.

Obviously a complete rewrite of the code base would mean that existing PRs to the master branch would be useless. (Not that I get any...)

I hope that the KWin 6.0 API stays stable, because it would suck to rewrite this code again. I do plan to support this code until a rewrite is necessary.

Edit - I meant a beta by the end of October/November, I wrote this message at the beginning of october

**Comment by dreamcat4** [comment]
@zeroxoneafour I have "pretty much" already written a draft design document for how to do underlying tiling mechanism constructs. And those base data structions to build upon. However I never shared it with nobody because I then thought: well, this document really needs some decent visualization diagrams!

So it took me 1 day to draw 1st diagram (using draio / draw . io desktop app). And then I got interrupted with it all.

Can I email you my design doc or something? Or gist it. And then you can look at the ideas being presented there.

Sorry I should add that after writing it all then saw what Cosmic had done. But they added some sweet extra stuff too (like tabs). Anyhow I have it on my disk.

Please register interest, as incomplete status. And then I will share as-is. Many thanks my friend for showing efforts for working on these features for KDE Desktop.

**Comment by zeroxoneafour** [comment]
@dreamcat4 I would be interested in seeing any ideas of how to do the tiling abstraction. I have a concept in my mind that I think would work, but any other ideas would be helpful, too.

**Comment by dreamcat4** [comment]
many thanks. although completed, my draft text was written a bit sloppy. so needs a little more time to be finished up well enough for readability sake. perhaps was now fixed during this evening the first 3-4 beginning sections... 

turns out the total full document was a lot longer than had remembered! some essays worth. so will keep going during next few days. with the aim to provide asap. very pleased for these renewed enthusiasm to finish it now! to have done then after some real recipient to pass it on to. to share these ideas out.

**Comment by gplusplus314** [comment]
I'm liking the renewed interest in this. Would you consider setting up a Discord server for the project?

**Comment by polair** [comment]
I would throw you another contribution for this, I'm not rich so it would be a small token but a clean function kwin-based tiling manager would be a dream. You have my kudos.

**Comment by Mithras** [comment]
I think reviewing bismuth code before starting anything new would be helpful.

**Comment by dreamcat4** [comment]
Quick update: For writing my general design / tiling inspirational type document - I am now (approx) 50% through redrafting / tidying process. However it' s still much like some "wall of text" or technical article. That the 1st hald of the document is mostly setup and "justifications" and some example use cases. So that has mostly been covered to where I am at. The 2nd half of the design document then increasingly brings into focus, and starts proposing relevant and necessary mechanisms, hierarchies and data structures. To tackle this (lets admit it) fairly complex problem space.

The end goal (by the end) is to just have some very general but minimal OO objects paradigm(s). For the "how to make frameworks skeletons in code". So not actually implementing specific features down to that level. So in those ways more architectural model.

BTW I should say it's a very generalized presentations. It does not comprehensively laying out exhaustively all of the possible and implementable full set of features. It just uses some of the more tricky ones as examples, to then guide and inform those more general "framework" type of overall design language and philosophy. Which then lays ground for something like prior lessesns learned. And some "hopefully" robust enough platform design concepts to builds upon. If that makes sense. Since actually implementing comprehensively an upgradable modern auto tiling solutions is so darned challenging.

SO another 50% remaining left to organise (into sections) tweak, (actually properly punctuate, fix the grammar), re-edit, and bring the document writing standard up. And become a navigable, orderly document. That is a standard enough that can then be read cleanly by 3rd parties.

**Comment by zeroxoneafour** [comment]
It's important to remember that this script is built on top of KWin's existing tiling and not around it, so it doesn't really share much similarity in the low level with Bismuth or most any other tiling manager that currently exists. That being said, I have heavily reviewed the Bismuth code and am trying to program in that style. The current code is in the v1.0 branch.

**Comment by gplusplus314** [comment]
@dreamcat4 - I'm happy to help with the docs, if you'd like.

**Comment by dreamcat4** [comment]
> @dreamcat4 - I'm happy to help with the docs, if you'd like.

it's really on me atm to get the draft up to scratch. however after releasing it then you will be more than welcome to make any further edits, revisions or changes however you wish. or reuse any relevant portions and snippets of the document.

so once released, you will be in a better position to see, and figure out how can tweak it. or improve further with your own spins / build upon. and there was actually already section written and completed (a short mini guide) all about recommendations for that. more than happy my friend.

**Comment by Thaodan** [comment]
> I'm liking the renewed interest in this. Would you consider setting up a Discord server for the project?

KDE Projects usually use Matrix/IRC. Discord would be less freedom.

**Comment by zeroxoneafour** [comment]
Bit of a roadblock, Kwin keeps throwing `TypeError: TypeError` when running the v1.0 branch code, despite the code having no issues during compilation, in the language server, or in eslint. There is also an error in a dependency that I used without a problem in the older version of Polonium. Thorough debugging and code analysis has found no issues, and all the tsconfigs and such were copied from the older script version. Truly a kwin scripting moment.

If anyone with more TS experience than me can try to debug v1.0 and find out the cause of the nameless TypeError in line 584 of the compiled `main.mjs`, I have an outline built for the rest of the script that I would like to build off of soon.

**Comment by zeroxoneafour** [comment]
v1.0 now works on an extremely basic level, probably around the level when kwin-autotile was very first released. If anyone wants to test it, it obviously wouldn't work as a replacement right now but it could show some promise.

**Comment by moutian** [comment]
> v1.0 now works on an extremely basic level, probably around the level when kwin-autotile was very first released. If anyone wants to test it, it obviously wouldn't work as a replacement right now but it could show some promise.

Thanks for your effort.  Could you post the compiled file as a alpha or beta release?  It would be much more convenient for us to try it.

**Comment by zeroxoneafour** [comment]
Github actions builds should now be working for the v1.0 branch.

**Comment by polair** [comment]
Heyo @zeroxoneafour. How would you prefer feedback on the "v1.0" build? Comments/issues here? I don't want to muddy the waters and make your life hard but I've been playing around.

**Comment by zeroxoneafour** [comment]
@polair If you find anything that has clearly been implemented and isn't working properly, a new issue would be preferable. This issue could work good for more general feedback.

**Comment by forlornidealist** [comment]
Thank you so much for your efforts! I will send you another sizeable donation soon 😊

**Comment by czhunor** [comment]
Did you set up any chat room or something where we would be able to contact you? I like the idea of tiling within KDE, I personally never used Bismuth, just saw a couple of videos, and heard that it is not really maintained anymore and should be rewritten. This is how I found out about your project, and would love to help with anything I can.

**Comment by zeroxoneafour** [comment]
@czhunor I've mainly been using Github for communication about the project, it seems like its been useful for keeping discussions focused. I have no interest in maintaining a larger server about the project.

**Comment by dreamcat4** [comment]
It seems that KDE project has now taken up matrix / element for chat services, so the thing to do would be:

1) join KDE's matrix space at #kde-community:kde.org
2) create new chat room there for these auto tiling specific discussions (if not already exists)

You can find here:

https://go.kde.org/matrix/#/#kde-community:kde.org

**Comment by czhunor** [comment]
That's a good idea, I would need a bit of help how to debug the script, I have never done anything with KDE development until now. I tried to set the flag for the debug log in the script and checked the output in the journalctl -f QT_CATEGORY=js QT_CATEGORY=kwin_scripting but i did not saw anything :)
I read the tutorial part of the scripting but i could not find anything about debugging there.

Is there any good tutorial to how to start with developing scripts for KDE?

I am a software developer for about 13 years now, but in completely different world (SAP) :)
